File: src/main/java/xyz/devvydont/smprpg/items/blueprints/resources/mining/MithrilFamilyBlueprint.java
package xyz.devvydont.smprpg.items.blueprints.resources.mining;

import xyz.devvydont.smprpg.items.CustomItemType;
import xyz.devvydont.smprpg.items.base.CustomCompressableBlueprint;
import xyz.devvydont.smprpg.services.ItemService;
import xyz.devvydont.smprpg.util.crafting.CompressionRecipeMember;
import xyz.devvydont.smprpg.util.crafting.MaterialWrapper;

import java.util.List;

public class MithrilFamilyBlueprint extends CustomCompressableBlueprint {

    public static final List<CompressionRecipeMember> COMPRESSION_FLOW = List.of(
            new CompressionRecipeMember(new MaterialWrapper(CustomItemType.MITHRIL_INGOT)),
            new CompressionRecipeMember(new MaterialWrapper(CustomItemType.MITHRIL_BLOCK)),
            new CompressionRecipeMember(new MaterialWrapper(CustomItemType.ENCHANTED_MITHRIL)),
            new CompressionRecipeMember(new MaterialWrapper(CustomItemType.ENCHANTED_MITHRIL_BLOCK)),
            new CompressionRecipeMember(new MaterialWrapper(CustomItemType.MITHRIL_SINGULARITY))
    );

    public MithrilFamilyBlueprint(ItemService itemService, CustomItemType type) {
        super(itemService, type);
    }

    @Override
    public List<CompressionRecipeMember> getCompressionFlow() {
        return COMPRESSION_FLOW;
    }

}

File: src/main/java/xyz/devvydont/smprpg/items/blueprints/resources/mining/MithrilFamilyBlueprint.kt
package xyz.devvydont.smprpg.items.blueprints.resources.mining

import xyz.devvydont.smprpg.items.CustomItemType
import xyz.devvydont.smprpg.items.ItemClassification
import xyz.devvydont.smprpg.items.base.CustomItemBlueprint
import xyz.devvydont.smprpg.items.interfaces.ICompressible
import xyz.devvydont.smprpg.items.interfaces.ICompressible.CompressionStep
import xyz.devvydont.smprpg.items.interfaces.ISellable
import xyz.devvydont.smprpg.services.ItemService
import xyz.devvydont.smprpg.util.extensions.calculateCompressedWorth
import org.bukkit.inventory.ItemStack

class MithrilFamilyBlueprint(itemService: ItemService, type: CustomItemType) :
    CustomItemBlueprint(itemService, type), ICompressible, ISellable {

    override val itemClassification get() = ItemClassification.MATERIAL

    override val decompressor: CompressionStep? get() = when (customItemType) {
        CustomItemType.MITHRIL_BLOCK          -> CompressionStep(itemService.getBlueprint(CustomItemType.MITHRIL_INGOT) as ICompressible, 1, 9)
        CustomItemType.ENCHANTED_MITHRIL      -> CompressionStep(itemService.getBlueprint(CustomItemType.MITHRIL_BLOCK) as ICompressible, 1, 9)
        CustomItemType.ENCHANTED_MITHRIL_BLOCK-> CompressionStep(itemService.getBlueprint(CustomItemType.ENCHANTED_MITHRIL) as ICompressible, 1, 9)
        CustomItemType.MITHRIL_SINGULARITY    -> CompressionStep(itemService.getBlueprint(CustomItemType.ENCHANTED_MITHRIL_BLOCK) as ICompressible, 1, 9)
        else -> null
    }

    override val compressor: CompressionStep? get() = when (customItemType) {
        CustomItemType.MITHRIL_INGOT          -> CompressionStep(itemService.getBlueprint(CustomItemType.MITHRIL_BLOCK) as ICompressible, 9, 1)
        CustomItemType.MITHRIL_BLOCK          -> CompressionStep(itemService.getBlueprint(CustomItemType.ENCHANTED_MITHRIL) as ICompressible, 9, 1)
        CustomItemType.ENCHANTED_MITHRIL      -> CompressionStep(itemService.getBlueprint(CustomItemType.ENCHANTED_MITHRIL_BLOCK) as ICompressible, 9, 1)
        CustomItemType.ENCHANTED_MITHRIL_BLOCK-> CompressionStep(itemService.getBlueprint(CustomItemType.MITHRIL_SINGULARITY) as ICompressible, 9, 1)
        else -> null
    }

    override fun getWorth(itemStack: ItemStack) = calculateCompressedWorth(itemStack)
}
